File: backend/src/core/Collection.ts
import { Static } from 'elysia';
import { bb, db } from './db';
import Field from './Field';
import { RuleData } from '../apis/handlers/collection/createCollection';
import { LengthRow } from '../db/models';

type CollectionProps = {
  id?: string;
  name: string;
  type: string;
  description?: string;
  fields: Field[];
  isSystem?: boolean;
  requiresAuth?: boolean;
  ruleData: Static<typeof RuleData>;
};

export default class Collection {
  id: string;
  name: string;
  type: string;
  description: string;
  fields: Field[];
  isSystem: boolean;
  requiresAuth: boolean;
  ruleData: Static<typeof RuleData>;

  idField = new Field({
    name: 'id',
    type: 'String',
    isSecure: false,
    isSystem: true,
    isHidden: false,
    isRequired: true,
    isPrimaryKey: true,
    isUnique: true,
  });

  constructor({
    id = '',
    name,
    type,
    description = '',
    fields,
    isSystem = false,
    requiresAuth = false,
    ruleData,
  }: CollectionProps) {
    this.name = name;
    this.type = type;
    this.description = description;

    this.fields = [this.idField, ...fields];

    this.isSystem = isSystem;
    this.requiresAuth = requiresAuth;

    this.ruleData = ruleData;

    this.id = id;

    if (id === '') {
      this.id = crypto.randomUUID();
    }
  }

  exists() {
    const result = bb
      .raw(
        `SELECT COUNT(*) AS length FROM collections_meta WHERE name = $name`,
        { $name: this.name }
      )
      .as(LengthRow)
      .get();

    if (result) {
      return result.length !== 0;
    }

    return false;
  }

  createTable() {
    let fieldsString = '';
    const primary_keys: string[] = [];

    for (const field of this.fields) {
      fieldsString += `${field.sql()},\n`;
      if (field.isPrimaryKey) {
        primary_keys.push(`${field.name}`);
      }
    }

    const createStatement = `
      CREATE TABLE IF NOT EXISTS ${this.name}(
          ${fieldsString}
          PRIMARY KEY(${primary_keys.join(', ')})
      )
  `;

    db.run(createStatement);
  }

  insertMetaData() {
    if (!this.exists()) {
      try {
        bb.insertInto('collections_meta')
          .values({
            id: this.id,
            name: this.name,
            type: this.type,
            description: this.description,
            requires_auth: this.requiresAuth,
            is_system: this.isSystem,
          })
          .run();

        for (const [key, value] of Object.entries(this.ruleData)) {
          bb.insertInto('auth_rules')
            .values({
              id: crypto.randomUUID(),
              collection: this.id,
              rule: key,
              permission: value,
            })
            .run();
        }

        for (const field of this.fields) {
          field.insertMetaData(this.id);
        }
      } catch (e) {
        console.log(e);
      }
    }
  }
}

File: Bird/Shared/Collection.cs
using System.Diagnostics.Eventing.Reader;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Authentication;
using SqlKata.Execution;



namespace Bird.Shared
{
    public record Length
    {
        public int count { get; init; }
    }

    public class Collection
    {
        private string id;
        private string name;
        private string description = string.Empty;
        private string relationAlias = string.Empty;
        private bool isSystem;
        private string type;
        private List<Field> fields;

        public Collection(string name, string type)
        {
            this.name = name;
            this.type = type;

            fields = [new Field(new FieldProps { Name = "id", Type = "String", IsRequired = true, IsPrimaryKey = true, IsUnique = true })];

            id = IdGenerator.GenerateShortUuid();
        }

        public void CreateTable(QueryFactory db)
        {
            var FieldTypes = new Dictionary<string, string>
                {
                    { "String", "text" },
                    { "Integer", "integer" },
                    { "Float", "float4" },
                    { "Boolean", "boolean" },
                    { "Date", "date" },
                    { "Select", "text" },
                    { "Relation", "text" },
                    { "File", "text" },
                    { "Markdown", "text" }
                };

            var columns = fields.Select(field =>
                $"{field.Name} {FieldTypes[field.Type]}" +
                (field.IsPrimaryKey ? " PRIMARY KEY" : "") +
                (field.IsRequired   ? " NOT NULL"    : "")
            );

            db.Statement($@"
                CREATE TABLE IF NOT EXISTS {name} (
                    {string.Join(", ", columns)}
                );
            ");

        }

        public async Task InsertMetaData(QueryFactory db)
        {
            if (!exists(db))
            {
                await db
                .Query("collections_meta")
                .InsertAsync(
                    new
                    {
                        id = id,
                        name = name,
                        type = type,
                        description = description,
                        is_system = isSystem,
                        relation_alias = relationAlias
                    }
                );

                foreach (var field in fields)
                {
                    await field.InsertMetaDataAsync(id, db);
                }
            }

        }

        private Boolean exists(QueryFactory db)
        {

            var result = db
                .Query("collections_meta")
                .Select("id")
                .AsCount()
                .Where("name", name)
                .First<Length>();

            return result.count != 0;
        }

        public Collection AddField(FieldProps props)
        {
            fields.Add(new Field(props));
            return this;
        }

        public Collection AddDescription(string description)
        {
            this.description = description;
            return this;
        }

        public Collection SetSystem(bool isSystem)
        {
            this.isSystem = isSystem;
            return this;
        }

        public Collection SetRuleData()
        {
            return this;
        }

        public Collection SetRelationAlias(string relationAlias)
        {
            this.relationAlias = relationAlias;
            return this;
        }
    }
}